Pushcollector accepts PushItem objects along with raw dict

rajulkumar opened the pull request:
Pushcollector library only accepted push items as raw dicts.
This patch modifies update_push_items to accept PushItem objects
from Pushsource library along with the raw dicts.

rajulkumar committed:
Pushcollector accepts PushItem objects along with raw dict

Pushcollector library only accepted push items as raw dicts.
This patch modifies update_push_items to accept PushItem objects
from Pushsource library along with the raw dicts.

rajulkumar committed:
Updated test-requirements and PushItem object initialization

PushItem obj can only be initialized with keyword args to init
for py3. Hence, updated the object init in tests.

rajulkumar committed:
Rpm package is installed for rpm-py-installer during travis env initialization

Rpm package should be installed for rpm-py-installer which is
required by pushsource library while testing. This rpm package
is installed via apt-get during travis environment initialization.

rohanpm commented on src/pushcollector/_impl/proxy.py:
```diff
@@ -50,11 +51,48 @@ class CollectorProxy(object):
+
+        return pushitem
+
+    def _is_pushitem_obj(self, pushitem):
```
Could we please not do it this way at all, i.e. can we please have no direct dependency on the specific class of the given object?

I suggest that this should use duck typing.  If it's a dict then use it as a dict, if it's not a dict then go via `_translate_pushitem` and let it crash if the object is of the wrong shape.  Then it accepts any objects which are compatible with the documented API of PushItem.

An example of a concrete issue caused here is that, by putting these requirements on the object's class and module, it prevents it from working with mocks.

rajulkumar replied:
+1
I wasn't sure if we can rely on the doc or we need strict type checking.

rohanpm commented on src/pushcollector/_impl/proxy.py:
```diff
@@ -50,11 +51,48 @@ class CollectorProxy(object):
     def update_push_items(self, items):
+        pushitems = []
         for item in items:
+            if self._is_pushitem_obj(item):
```
This is the part which I would suggest changing.  Actually how about writing it like this:

```suggestion
            item_dicts = self._translate_pushitem(item)
```

`_translate_pushitem` itself can have logic "if input is a dict, just return [item], else convert it to dicts of the expected format".

rohanpm commented on tests/proxy/test_argument_handling.py:
```diff
@@ -12,3 +15,58 @@ def test_bad_push_items():
+    assert ret_val.result() is None
+
+
+def test_bad_obj_push():
```
I don't think we should do this... just accept objects if they have the valid attributes.

rohanpm commented on src/pushcollector/_impl/proxy.py:
```diff
@@ -50,11 +51,48 @@ class CollectorProxy(object):
+            "filename": pushitem.name,
+            "state": pushitem.state,
+            "src": pushitem.src,
+            "dest": str(pushitem.dest) if pushitem.dest else None,
```
This doesn't look correct to me, two problems:

- if it's right to serialize a list into a string like this, I don't think python str() would be the right way... I don't think there's any guarantee about that format.
- it won't give a correct result anyway to collapse the list of destinations into a single string here.  The way the push items are stored in Pub today is for this field to end up containing a single destination only.  If you have a look in `log_push_items` method in Pub, you can see there's a loops where it takes input items which may contain a *list* of destinations, and ends up producing one output push item for each destination on each input item.  I think the same logic is needed here.

So it seems like `_translate_pushitem(self, pushitem)` should be defined as returning a *list of* dicts rather than just one dict.

rajulkumar replied:
I wasn't sure on this part whether there should be a push item for each destination or one push item with the destination as it's just collecting the push info like logs. Probably should have put a comment here earlier.

rohanpm commented on src/pushcollector/_impl/proxy.py:
```diff
@@ -50,11 +51,48 @@ class CollectorProxy(object):
         self._delegate = delegate
 
+    def _translate_pushitem(self, pushitem):
+        def _get_checksum():
```
The style here seems pretty odd, wouldn't it be less code and more straightforward if this was simply a variable assignment?

```
  checksums = None
  if (...):
    checksums = { ... }
```

I'm not getting an advantage from wrapping it up into a closure.

rajulkumar replied:
+1

rohanpm commented on src/pushcollector/_impl/proxy.py:
```diff
@@ -50,11 +51,48 @@ class CollectorProxy(object):
 
+    def _translate_pushitem(self, pushitem):
+        def _get_checksum():
+            if pushitem.md5sum and pushitem.sha256sum:
```
Why do we only save checksums if *both* of these are available?

rajulkumar replied:
I understood the json schema incorrectly where it looked like either the checksums is null or has both the checksums. Also checking on Pub, it looked like that's how it is.

rohanpm commented on tests/proxy/test_argument_handling.py:
```diff
@@ -12,3 +15,58 @@ def test_bad_push_items():
+    )
+    collector.update_push_items([pushitem])
+
+    update_push_item_args = collector._delegate.update_push_items.call_args[0][0]
```
This test relies on the implementation detail that get() returns an object with some _delegate pointing at the backend object.  This is unrelated to what you want to test.  Can we please avoid unnecessary dependencies on details like that?  It increases the cost of refactoring.

I don't see any reason the test needs to be written in this way.  Your test right now sets up the backend to create a new Mock object whenever an instance is requested, and you then need to rely on some implementation details to get a reference to that object.  Why not just register your backend to return a mock for which you already hold a reference:

```
mock = Mock()
# This backend always returns the mock we prepared.
Collector.register_backend("mock", lambda: mock)
collector = Collector.get("mock")

...

update_push_item_args = mock.update_push_items...
```

rajulkumar replied:
+1

rohanpm commented on src/pushcollector/_impl/collector.py:
```diff
@@ -20,8 +20,9 @@ def update_push_items(self, items):
         Arguments:
-            items (List[dict])
-                    A list of push item dicts.
+            items (List[dict], List[:class:`~pushsource.PushItem`])
```
Have you checked if this renders correctly when docs are generated?

In order for the pushsource.PushItem class to be resolved and hyperlinked correctly, I think you would typically need to add an entry to intersphinx_mapping in docs/conf.py.

rajulkumar replied:
It compiled properly using make but the link to pushsouce.PushItem didn't work as expected.
However, there was an error rendering this in tox environment.
Updating intersphinx_mapping works as expected.

rajulkumar committed:
Ducktyping instead of strict type checking and pushitem for each destination

The pushitem objects are not strictly checked for a specifc class
type. Instead they are expected to have requried attributes. Also,
pushitems are geenrated for each destination in the object.

rajulkumar committed:
Code formatting via black

rajulkumar committed:
Static code checks are successfully completed

rohanpm commented on src/pushcollector/_impl/collector.py:
```diff
@@ -20,8 +20,9 @@ def update_push_items(self, items):
-                    A list of push item dicts.
+            items (List[dict], List[:class:`~pushsource.PushItem`])
+                    A list of push item dicts or list of :class:`~pushsource.PushItem`
+                    objects
```
```suggestion
                    objects.
```

rohanpm commented on src/pushcollector/_impl/proxy.py:
```diff
@@ -50,11 +50,45 @@ class CollectorProxy(object):
+            "signing_key": pushitem.signing_key,
+        }
+
+        if pushitem.dest:
+            for dest in pushitem.dest:
```
This is just a stylistic preference, pointing it out in case you haven't considered it...

I'd have written this instead as:

```suggestion
        for dest in (pushitem.dest or []):
```

rohanpm commented on src/pushcollector/_impl/proxy.py:
```diff
@@ -50,11 +50,45 @@ class CollectorProxy(object):
+                push_item_copy = push_item.copy()
+                push_item_copy["dest"] = dest
+                pushitems.append(push_item_copy)
+
```
The logic here all looks right, probably could have benefited from a comment or two though.

We've just had a discussion in code review so it's fresh in our minds, but for someone reading this in a different context it doesn't seem obvious that the intended behavior here is:

- If object has multiple dests, we return one dict per dest.
- If object has no dest at all, we return one dict with an empty dest.

Without a comment mentioning this then one might start to wonder e.g. "is this really intended - why not return an empty list instead for no dest?"

rajulkumar committed:
Adds test for minimal PushItem object and comments in proxy.py
